Add alias filtering to invoke subgraph

tugsbayasgalan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #169134


invoke_subgraph is NOT desugared or inlined by AOTDispatcher. it remains as an opaque HOP call in the final graph
The HOP infrastructure assumes subgraphs have no input-output or output-output aliasing. This becomes problematic when you return intermediates as HOP outputs. This PR adds checks to make sure these intermediates don't alias the actual inp/output.  To demonstrate what i mean, consider following example:
```
@invoke_subgraph
  def gn(x):
      view = x.view(2, 4)  # intermediate that aliases input x
      y = torch.sin(view)
      return torch.cos(view)

def fn(x):
       res = gn(x)
       return res + 4 
```

In this case, if we don't filter `y` we would later error because some HOP has strict aliasing check on input/outputs. This does however introduce a subtle issue when we do something like:
```
@invoke_subgraph
  def gn(x):
      view = x.view(2, 4)  # intermediate that aliases input x
      y = torch.sin(view) 
      captured.append(y)
      return torch.cos(view)

captured = [] 
def fn(x):
       res = gn(x)
       return res + 4 
```
because in this case, we will not replay side effect on `captured` in the graph which fails with not a nice error. We will address it in follow up PR because i think this case is rare. This is not a regression because side effect was never supported for invoke_subgraph anyways. 


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @Lucaskabela @jataylo

tugsbayasgalan committed:
Add alias filtering to invoke subgraph

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Add alias filtering to invoke subgraph"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Add alias filtering to invoke subgraph"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Add alias filtering to invoke subgraph"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Add alias filtering to invoke subgraph"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Add alias filtering to invoke subgraph"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Add alias filtering to invoke subgraph"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Add alias filtering to invoke subgraph"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Add alias filtering to invoke subgraph"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Add alias filtering to invoke subgraph"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Add alias filtering to invoke subgraph"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

zou3519 reviewed (commented):
can you explain what alias filtering is? there's no description in this PR

tugsbayasgalan committed:
Update on "Add alias filtering to invoke subgraph"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Add alias filtering to invoke subgraph"



invoke_subgraph is NOT desugared or inlined by AOTDispatcher. it remains as an opaque HOP call in the final graph
The HOP infrastructure assumes subgraphs have no input-output or output-output aliasing. This becomes problematic when you return intermediates as HOP outputs. This PR adds checks to make sure these intermediates don't alias the actual inp/output. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Add alias filtering to invoke subgraph"



invoke_subgraph is NOT desugared or inlined by AOTDispatcher. it remains as an opaque HOP call in the final graph
The HOP infrastructure assumes subgraphs have no input-output or output-output aliasing. This becomes problematic when you return intermediates as HOP outputs. This PR adds checks to make sure these intermediates don't alias the actual inp/output. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

anijain2305 reviewed (commented):
Can you take a look at `has_aliasing` in output_graph.py and share the code? Its not directly usable, but I want to avoid code duplication.

tugsbayasgalan committed:
Update on "Add alias filtering to invoke subgraph"



invoke_subgraph is NOT desugared or inlined by AOTDispatcher. it remains as an opaque HOP call in the final graph
The HOP infrastructure assumes subgraphs have no input-output or output-output aliasing. This becomes problematic when you return intermediates as HOP outputs. This PR adds checks to make sure these intermediates don't alias the actual inp/output. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Add alias filtering to invoke subgraph"



invoke_subgraph is NOT desugared or inlined by AOTDispatcher. it remains as an opaque HOP call in the final graph
The HOP infrastructure assumes subgraphs have no input-output or output-output aliasing. This becomes problematic when you return intermediates as HOP outputs. This PR adds checks to make sure these intermediates don't alias the actual inp/output. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

anijain2305 reviewed (approved):
LGTM with few comments. But can you have @zou3519  review as well?

tugsbayasgalan committed:
Update on "Add alias filtering to invoke subgraph"



invoke_subgraph is NOT desugared or inlined by AOTDispatcher. it remains as an opaque HOP call in the final graph
The HOP infrastructure assumes subgraphs have no input-output or output-output aliasing. This becomes problematic when you return intermediates as HOP outputs. This PR adds checks to make sure these intermediates don't alias the actual inp/output. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Add alias filtering to invoke subgraph"



invoke_subgraph is NOT desugared or inlined by AOTDispatcher. it remains as an opaque HOP call in the final graph
The HOP infrastructure assumes subgraphs have no input-output or output-output aliasing. This becomes problematic when you return intermediates as HOP outputs. This PR adds checks to make sure these intermediates don't alias the actual inp/output.  To demonstrate what i mean, consider following example:
```
invoke_subgraph
  def gn(x):
      view = x.view(2, 4)  # intermediate that aliases input x
      y = torch.sin(view)
      return torch.cos(view)

def fn(x):
       res = gn(x)
       return res + 4 
```

In this case, if we don't filter `y` we would later error because some HOP has strict aliasing check on input/outputs. This does however introduce a subtle issue when we do something like:
```
invoke_subgraph
  def gn(x):
      view = x.view(2, 4)  # intermediate that aliases input x
      y = torch.sin(view) 
      captured.append(y)
      return torch.cos(view)

captured = [] 
def fn(x):
       res = gn(x)
       return res + 4 
```
because in this case, we will not replay side effect on `captured` in the graph which fails with not a nice error. We will address it in follow up PR because i think this case is rare. This is not a regression because side effect was never supported for invoke_subgraph anyways. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

tugsbayasgalan commented on torch/_dynamo/variables/higher_order_ops.py:
```diff
@@ -4262,8 +4389,9 @@ def _call_function(
-    # TODO - make this true to support mutation
-    allow_side_effects = False
+    allow_side_effects = True
+    # invoke_subgraph is NOT desugared, so we need strict aliasing filtering
```
updated

zou3519 commented on torch/_dynamo/variables/higher_order_ops.py:
```diff
@@ -393,15 +393,127 @@ def _assert_tensors_nonaliasing(inputs, outputs):
+
+    def check_and_track(self, proxy_node) -> bool:
+        """
+        Check if a proxy node can be added without aliasing, and track it if so.
```
I dont know what "if a proxy node can be added without aliasing" means. Added to what without aliasing? Why are we tracking proxy nodes? I thought we were tracking storages.

zou3519 commented on torch/_dynamo/variables/higher_order_ops.py:
```diff
@@ -393,15 +393,127 @@ def _assert_tensors_nonaliasing(inputs, outputs):
+        get_plain_tensors(tensor, inner_tensors)
+        for inner in inner_tensors:
+            if isinstance(inner, torch.Tensor):
+                storages.add(StorageWeakRef(inner._typed_storage()))
```
I would really just simplify this function right now by not worrying about the tensor subclass case. invoke_subgraph doesn't support tensor subclasses. so your tests are just going to raise an error.

zou3519 reviewed (approved):
would recommend just simplifying this to not use tensor subclasses right now because I don't think invoke_subgraph supports tensor subclasses. Also, added more suggestions for code comments that would have clarified the behavior for me. otherwise lgtm

tugsbayasgalan committed:
Update on "Add alias filtering to invoke subgraph"



invoke_subgraph is NOT desugared or inlined by AOTDispatcher. it remains as an opaque HOP call in the final graph
The HOP infrastructure assumes subgraphs have no input-output or output-output aliasing. This becomes problematic when you return intermediates as HOP outputs. This PR adds checks to make sure these intermediates don't alias the actual inp/output.  To demonstrate what i mean, consider following example:
```
invoke_subgraph
  def gn(x):
      view = x.view(2, 4)  # intermediate that aliases input x
      y = torch.sin(view)
      return torch.cos(view)

def fn(x):
       res = gn(x)
       return res + 4 
```

In this case, if we don't filter `y` we would later error because some HOP has strict aliasing check on input/outputs. This does however introduce a subtle issue when we do something like:
```
invoke_subgraph
  def gn(x):
      view = x.view(2, 4)  # intermediate that aliases input x
      y = torch.sin(view) 
      captured.append(y)
      return torch.cos(view)

captured = [] 
def fn(x):
       res = gn(x)
       return res + 4 
```
because in this case, we will not replay side effect on `captured` in the graph which fails with not a nice error. We will address it in follow up PR because i think this case is rare. This is not a regression because side effect was never supported for invoke_subgraph anyways. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Add alias filtering to invoke subgraph"



invoke_subgraph is NOT desugared or inlined by AOTDispatcher. it remains as an opaque HOP call in the final graph
The HOP infrastructure assumes subgraphs have no input-output or output-output aliasing. This becomes problematic when you return intermediates as HOP outputs. This PR adds checks to make sure these intermediates don't alias the actual inp/output.  To demonstrate what i mean, consider following example:
```
invoke_subgraph
  def gn(x):
      view = x.view(2, 4)  # intermediate that aliases input x
      y = torch.sin(view)
      return torch.cos(view)

def fn(x):
       res = gn(x)
       return res + 4 
```

In this case, if we don't filter `y` we would later error because some HOP has strict aliasing check on input/outputs. This does however introduce a subtle issue when we do something like:
```
invoke_subgraph
  def gn(x):
      view = x.view(2, 4)  # intermediate that aliases input x
      y = torch.sin(view) 
      captured.append(y)
      return torch.cos(view)

captured = [] 
def fn(x):
       res = gn(x)
       return res + 4 
```
because in this case, we will not replay side effect on `captured` in the graph which fails with not a nice error. We will address it in follow up PR because i think this case is rare. This is not a regression because side effect was never supported for invoke_subgraph anyways. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Add alias filtering to invoke subgraph"



invoke_subgraph is NOT desugared or inlined by AOTDispatcher. it remains as an opaque HOP call in the final graph
The HOP infrastructure assumes subgraphs have no input-output or output-output aliasing. This becomes problematic when you return intermediates as HOP outputs. This PR adds checks to make sure these intermediates don't alias the actual inp/output.  To demonstrate what i mean, consider following example:
```
invoke_subgraph
  def gn(x):
      view = x.view(2, 4)  # intermediate that aliases input x
      y = torch.sin(view)
      return torch.cos(view)

def fn(x):
       res = gn(x)
       return res + 4 
```

In this case, if we don't filter `y` we would later error because some HOP has strict aliasing check on input/outputs. This does however introduce a subtle issue when we do something like:
```
invoke_subgraph
  def gn(x):
      view = x.view(2, 4)  # intermediate that aliases input x
      y = torch.sin(view) 
      captured.append(y)
      return torch.cos(view)

captured = [] 
def fn(x):
       res = gn(x)
       return res + 4 
```
because in this case, we will not replay side effect on `captured` in the graph which fails with not a nice error. We will address it in follow up PR because i think this case is rare. This is not a regression because side effect was never supported for invoke_subgraph anyways. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

tugsbayasgalan commented:
@pytorchbot merge -i
